backend/danswer/search/pipeline.py

```diff
@@ -48,15 +77,148 @@ def __init__(
+        self._reranked_chunks: list[InferenceChunk] | None = None
+        self._reranked_sections: list[InferenceSection] | None = None
+        self._relevant_chunk_indices: list[int] | None = None
+
+        # If chunks have been merged, the LLM filter flow no longer applies
+        # as the indices no longer match. Can be implemented later as needed
+        self.ran_merge_chunk = False
 
         # generator state
         self._postprocessing_generator: Generator[
             list[InferenceChunk] | list[str], None, None
         ] | None = None
 
+    def _combine_chunks(self, post_rerank: bool) -> list[InferenceSection]:
+        if not post_rerank and self._retrieved_sections:
+            return self._retrieved_sections
+        if post_rerank and self._reranked_sections:
+            return self._reranked_sections
+
+        if not post_rerank:
+            chunks = self.retrieved_chunks
+        else:
+            chunks = self.reranked_chunks
+
+        if self._search_query is None:
+            # Should never happen
+            raise RuntimeError("Failed in Query Preprocessing")
+
+        functions_with_args: list[tuple[Callable, tuple]] = []
+        final_inference_sections = []
+
+        # Nothing to combine, just return the chunks
+        if (
+            not self._search_query.chunks_above
+            and not self._search_query.chunks_below
+            and not self._search_query.full_doc
+        ):
+            return [InferenceSection.from_chunk(chunk) for chunk in chunks]
+
+        # If chunk merges have been run, LLM reranking loses meaning
+        # Needs reimplementation, out of scope for now
+        self.ran_merge_chunk = True
+
+        # Full doc setting takes priority
+        if self._search_query.full_doc:
+            seen_document_ids = set()
+            unique_chunks = []
+            for chunk in chunks:
+                if chunk.document_id not in seen_document_ids:
+                    seen_document_ids.add(chunk.document_id)
+                    unique_chunks.append(chunk)
+
+                    functions_with_args.append(
+                        (
+                            self.document_index.id_based_retrieval,
+                            (
+                                chunk.document_id,
+                                None,  # Start chunk ind
+                                None,  # End chunk ind
+                                # There is no chunk level permissioning, this expansion around chunks
+                                # can be assumed to be safe
+                                IndexFilters(access_control_list=None),
+                            ),
+                        )
+                    )
+
+            list_inference_chunks = run_functions_tuples_in_parallel(
+                functions_with_args, allow_failures=False
+            )
+
+            for chunk in unique_chunks:
+                inf_chunks = list_inference_chunks.pop(0)
```

Comment:
same comment here

Reply:
The ordering must be maintained for this to be correct, we could do everything from the back potentially but, I don't want to complicate it, I'll just use indices instead.

But also the scale of this is fairly small, shouldn't matter no?